### DISCUSSION TITLE
[LunarVim] Trouble with warning diagnostic messages

### DISCUSSION THREAD

**Discussion by frolvanya** [mrcjkb/rustaceanvim#417] [Q&A]
Recently I started experiencing strange warning messages behaviour
1) Warning message doesn't show up until I save file (if I remember correctly, few weeks before it showed up once I enter mormal mode). However, error messages work like a charm, instantly when I press Esc.
2) In some projects I don't see warning messages at all (I've tried running `cargo clean`)

https://github.com/mrcjkb/rustaceanvim/assets/59515280/8d717805-a61e-49d2-9db9-86624237115d

System: macOS Sonoma 14.5
Neovim: 0.10 stable
Lunarvim: release-1.4
Rustaceanvim: 4.24.0
Rustanalyzer: 2024-06-03

Rust + trouble setup:
```lua
{
    "mrcjkb/rustaceanvim",
    ft = { "rust" },
    config = function()
        vim.g.rustaceanvim = {
            tools = {
                float_win_config = {
                    border = "rounded"
                }
            },
            server = {
                on_attach = require("lvim.lsp").common_on_attach
            },
        }
    end,
},
{
    "saecki/crates.nvim",
    version = "v0.4.0",
    dependencies = { "nvim-lua/plenary.nvim" },
    config = function()
        require("crates").setup {
            null_ls = {
                enabled = true,
                name = "crates.nvim",
            },
            popup = {
                border = "rounded",
            },
        }
    end,
},
{
    "folke/trouble.nvim",
    opts = {},
    cmd = "Trouble",
}
```

What I am doing wrong? Recently I reinstalled lunarvim as a nix package + few days ago trouble.nvim v3 came out. Can something from this cause this problem?

EDIT: I've disabled trouble, but the issue persists + I've noticed that it stoped working on a somewhat big projects (20k LoC, 500+ deps)

EDIT2: Seems like simple laptop reboot solved all issues. However, it's strange when Zed is working just fine and nvim just dies for no reason

**Comment by mrcjkb** [comment]
> EDIT2: Seems like simple laptop reboot solved all issues. However, it's strange when Zed is working just fine and nvim just dies for no reason

Hey 👋 
So this is resolved or are you still facing issues?

[This](https://github.com/mrcjkb/rustaceanvim/issues/312) may be related to your issues with large projects.
See also: https://github.com/neovim/neovim/issues/23291

**Reply by frolvanya** [reply]
> So this is resolved or are you still facing issues?

Hi, yep, it's resolved by simple restart

Thanks for sharing those issues!

**Comment by frolvanya** [comment]
@mrcjkb Hmm, it seems like I have this issue again in some repos (paradoxically it doesn't happen on a big project now, but I can see this problem in quite simple project. It doesn't spike my cpu and again Zed works fine. Do you have any other ideas why it can happen?

<img width="1215" alt="image" src="https://github.com/mrcjkb/rustaceanvim/assets/59515280/d798bf47-463f-4606-b840-d309de9f6a16">

![image](https://github.com/mrcjkb/rustaceanvim/assets/59515280/c09cb1ae-09a0-4d3a-817a-71bd71385e0f)
![image](https://github.com/mrcjkb/rustaceanvim/assets/59515280/a64c75f2-8385-4633-89df-b1816bb2abc2)
![image](https://github.com/mrcjkb/rustaceanvim/assets/59515280/d33ec0cb-cda6-4de2-86a5-91c0ef2c8afc)

**Reply by mrcjkb** [reply]
I couldn't think of anything off the top of my head.
If you can reproduce it, I would suggest following the readme's troubleshooting section, starting with a minimal config and then bisecting your plugins.

**Reply by frolvanya** [reply]
Okay, I have absolutely no idea why it fixed an issue, but I'm just glad that it helped. For those struggling with inconsistent error messages, try to remove ftplugin/rust.lua. I used this file to activate some keybindings that are only for rust, nothing special was done there. I've deleted it and moved all mappings to the global file and everything started working:
```lua
lvim.builtin.which_key.mappings["r"] = {
    name = "Rust",
    r = { "<cmd>RustLsp runnables<cr>", "Runnables" },
    t = { "<cmd>RustLsp testables<cr>", "Testables" },
    m = { "<cmd>RustLsp expandMacro<cr>", "Expand Macro" },
    a = { "<cmd>RustLsp codeAction<cr>", "Code Actions" },
    h = { "<cmd>RustLsp hover actions<cr>", "Hover" },
    e = { "<cmd>RustLsp explainError<cr>", "Explain Error" },
    D = { "<cmd>RustLsp renderDiagnostic<cr>", "Render Diagnostic" },
    c = { "<cmd>RustLsp openCargo<cr>", "Open Cargo.toml" },
    d = { "<cmd>RustLsp openDocs<cr>", "Open Docs.rs" },
}

local crates = require("crates")
lvim.builtin.which_key.mappings["c"] = {
    name = "Crates",
    t = { crates.toggle, "Toogle" },
    r = { crates.reload, "Reload" },
    v = { crates.show_versions_popup, "Show Versions" },
    f = { crates.show_features_popup, "Show Features" },
    d = { crates.show_dependencies_popup, "Show Dependencies" },
    u = { crates.upgrade_crate, "Upgrade Crate" },
    a = { crates.upgrade_all_crates, "Update All Crates" },
    e = { crates.expand_plain_crate_to_inline_table, "Expand Crate" },
    E = { crates.extract_crate_into_table, "Extract Crate" },
    H = { crates.open_homepage, "Open Homepage" },
    R = { crates.open_repository, "Open Repository" },
    D = { crates.open_documentation, "Open Documentation" },
    C = { crates.open_crates_io, "Open Crates.io" },
}
```

**Comment by frolvanya** [comment]
@mrcjkb Sorry for bothering you again( Last fix didn't help (it helped only for some projects and I thought that it completely vanished my issue, so I was just lucky). However, I found out that it's a rustaceanvim failure:
- I'm using LunarVim and I need to manually disable rust-analyzer in LSP config to avoid duplication of all messages. Here's what happens when I allow two rust-analyzers work, as you can see only warning message wasn't duplicated:
<img width="1377" alt="image" src="https://github.com/mrcjkb/rustaceanvim/assets/59515280/149926b9-8eb0-42fa-ae1e-7c4ae14ea9a0">

- Rust-tools works perfectly fine:
<img width="854" alt="image" src="https://github.com/mrcjkb/rustaceanvim/assets/59515280/6c571eab-a4a0-45f1-b76e-6d9d685e7c1f">

Here's what happens when I use rustaceanvim (it's worth to mention that some projects work fine, maybe something is wrong with this [one](https://github.com/frolvanya/powerline-rs/tree/feat/moved-to-24-bits)):
<img width="852" alt="image" src="https://github.com/mrcjkb/rustaceanvim/assets/59515280/56923111-8698-4130-8806-56cd8c45c1ee">

Maybe I'm configuring it wrong? Can you please say if this is a correct configuration for LunarVim (I took this config from this [comment](https://github.com/mrcjkb/rustaceanvim/discussions/174#discussioncomment-8193827))

`plugins.lua`
```lua
{
    "mrcjkb/rustaceanvim",
    ft = { "rust" },
    config = function()
        vim.g.rustaceanvim = {
            tools = {
                float_win_config = {
                    border = "rounded"
                }
            },
            server = {
                on_attach = require("lvim.lsp").common_on_attach
            },
        }
    end,
}
```

`lsp.lua`
```lua
vim.list_extend(lvim.lsp.automatic_configuration.skipped_servers, { "clangd", "rust_analyzer" })
```

Will it help you to fix an issue if I'd try it out on raw neovim?

**Reply by mrcjkb** [reply]
Yes, this plugin conflicts with nvim-lspconfig: https://github.com/mrcjkb/rustaceanvim/discussions/174

**Reply by frolvanya** [reply]
Wdym? I already know this. I said that I enabled duplication **on purpose** to check whether it's a rustaceanvim problem in first bullet point. Everything was duplicated except warnings (that means that rustaceanvim's rust-analyzer didn't work properly). After that I tried rust-tools and everything worked as expected. However, when I'm using rustaceanvim + disabling second rust-analyzer, I don't have diagnostic messages (in some projects). Moreover, as I said, I'm using the "fix" from the discussion that you provided (see my last two files, plugins.lua and lsp.lua)

EDIT: Rust-tools works without any issues, so I'll stick to it instead of rustaceanvim. I guess we can close this discussion. Thanks for all answers!

**Reply by mrcjkb** [reply]
Yep, as I mentioned in the other discussion, I'm afraid I don't have the capacity to support every neovim distro out there. 
So if it's not reproducible with plain neovim and a minimal config (see the readme's troubleshooting section and the bug report issue template), there's not much I can do to help.
